File: python/lombokalgoritma/geometry/bezier.py
# LombokAlgoritma — Bézier curve evaluation, de Casteljau (SPEC §13.2)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from collections.abc import Sequence

from ..errors import EmptyInputError
from .types import Point2D


def bezier(control_points: Sequence[Point2D], t: float) -> Point2D:
    """Point of the Bézier curve at t: each level p' = p·s + q·t with s = 1 − t.

    Raises:
        EmptyInputError: no control points.
    """
    if len(control_points) == 0:
        raise EmptyInputError("bezier (need >= 1 control point)")
    pts: list[Point2D] = [(p[0], p[1]) for p in control_points]
    s = 1 - t
    while len(pts) > 1:
        pts = [
            (p[0] * s + q[0] * t, p[1] * s + q[1] * t) for p, q in zip(pts, pts[1:], strict=False)
        ]
    return pts[0]

File: typescript/src/geometry/bezier.ts
// LombokAlgoritma — Bézier curve evaluation (de Casteljau)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { EmptyInputError } from '../core/errors.js';
import type { Point2D } from './types.js';

/**
 * Point of the Bézier curve with the given control points at parameter t, by de Casteljau:
 * each level replaces pᵢ with `pᵢ·(1 − t) + pᵢ₊₁·t` (SPEC §7.5). O(n²).
 *
 * @throws RangeError with no control points
 */
export function bezier(controlPoints: readonly Point2D[], t: number): Point2D {
  if (controlPoints.length === 0) throw new EmptyInputError('bezier (need ≥ 1 control point)');
  let pts = controlPoints.map((p) => ({ x: p.x, y: p.y }));
  const s = 1 - t;
  while (pts.length > 1) {
    const next: Point2D[] = [];
    for (let i = 0; i + 1 < pts.length; i++) {
      const p = pts[i] as Point2D;
      const q = pts[i + 1] as Point2D;
      next.push({ x: p.x * s + q.x * t, y: p.y * s + q.y * t });
    }
    pts = next;
  }
  return pts[0] as Point2D;
}
